services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/application/ApplicationServiceTest.kt

```diff
@@ -65,37 +68,58 @@ import org.springframework.test.context.junit.jupiter.SpringExtension
 private const val USERNAME = "test"
 private const val HANKE_TUNNUS = "HAI-1234"
 
-@TestInstance(TestInstance.Lifecycle.PER_CLASS)
 @ExtendWith(SpringExtension::class)
 class ApplicationServiceTest {
-    private val applicationRepo: ApplicationRepository = mockk()
-    private val statusRepo: AlluStatusRepository = mockk()
-    private val cableReportService: CableReportService = mockk()
-    private val geometriatDao: GeometriatDao = mockk()
-    private val disclosureLogService: DisclosureLogService = mockk(relaxUnitFun = true)
-    private val applicationLoggingService: ApplicationLoggingService = mockk(relaxUnitFun = true)
-    private val hankeRepository: HankeRepository = mockk()
-    private val permissionService: PermissionService = mockk()
-    private val emailSenderService: EmailSenderService = mockk()
-    private val hankeKayttajaService: HankeKayttajaService = mockk(relaxUnitFun = true)
-    private val attachmentService: ApplicationAttachmentService = mockk()
-    private val hankeLoggingService: HankeLoggingService = mockk(relaxUnitFun = true)
-
-    private val applicationService: ApplicationService =
-        ApplicationService(
-            applicationRepo,
-            statusRepo,
-            cableReportService,
-            disclosureLogService,
-            applicationLoggingService,
-            hankeKayttajaService,
-            emailSenderService,
-            attachmentService,
-            geometriatDao,
-            permissionService,
-            hankeRepository,
-            hankeLoggingService,
-        )
+
+    @MockK private lateinit var applicationRepository: ApplicationRepository
+    @MockK private lateinit var hankeRepository: HankeRepository
+    @MockK private lateinit var statusRepository: AlluStatusRepository
+    @MockK private lateinit var geometriatDao: GeometriatDao
+
+    @MockK private lateinit var cableReportService: CableReportService
+    @MockK private lateinit var permissionService: PermissionService
+    @MockK private lateinit var emailSenderService: EmailSenderService
+    @MockK private lateinit var attachmentService: ApplicationAttachmentService
+
+    @MockK(relaxUnitFun = true) private lateinit var disclosureLogService: DisclosureLogService
+    @MockK(relaxUnitFun = true) private lateinit var loggingService: ApplicationLoggingService
+    @MockK(relaxUnitFun = true) private lateinit var hankeKayttajaService: HankeKayttajaService
+    @MockK(relaxUnitFun = true) private lateinit var hankeLoggingService: HankeLoggingService
+    @MockK(relaxed = true) private lateinit var featureFlags: FeatureFlags
+
+    @InjectMockKs private lateinit var applicationService: ApplicationService
```

Comment:
Why would you replace a concrete constructor with some kind of magick?

When ApplicationService dependencies change, this will give runtime errors for missing mocks. Or this might leave unnecessary mocks some dependencies are removed. Using a concrete constructor gives compile errors if the dependencies don't match, which is easier to spot. Also, it's clear which mocks are used and which are not.

Reply:
I thought annotations would have been a simpler way to initialise the service (ioc). If there is a missing mock, I don't think this will compile. But yea, this is somewhat ugly and and the previous style was maybe more readable. I'll revert.